### DISCUSSION TITLE
My docker image for YOLO deployment to Jetson nano

### DISCUSSION THREAD

**Discussion by kyuhyong** [ultralytics/ultralytics#19613] [Solutions]
Hello, 
I am interested in deploying AI on edge devices like jetson nano and have been trying to figure out how to best optimize the performace leveraging its gpu capability.

Installing ultralytics on fresh install from nvidia's official SD card image requires following steps.
- Install python3.8 with venv
- Build install torch from source v1.8.0 with gpu enabled (This will generate a wheel file)
- Build install torchvision from source v0.9.0
- Build install tensorrt python binding for best optimization.
- Build install ultralytics specifying exact version of **torch** and **torchvision**.

The result is here.
https://youtu.be/nxf9dNMFcmI

I also made a docker image just build for running ultralytics and you can simply run it on your jetson nano.
`docker pull kyubot/jetson_yolo`

Please note, this image is based from nvcr.io/nvidia/l4t-base:r32.7.1

Before running the docker, you will need to apt upgrade your board by `apt upgrade` command.
Thanks,

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kyuhyong, thank you for sharing your work and interest in deploying Ultralytics 🚀 on edge devices like the Jetson Nano! Your contribution and detailed steps will surely be valuable to others in the community working on similar projects.

For users exploring deployment on Jetson Nano or other customized environments, we recommend ensuring compatibility with the latest `ultralytics` package. Please provide additional details if you encounter any challenges or issues, such as specific error messages, logs, or a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) 🐛 to help us debug.

Additionally, if you're optimizing performance, you may find our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) and deployment guides useful. For Docker users, we encourage sharing your Dockerfile configuration if you'd like feedback or suggestions for improvements.

## Upgrade

Ensure you're using the latest `ultralytics` package by upgrading your environment. In a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) installed, run:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in various environments. While you've already tailored a Docker image for Jetson Nano, others can explore these verified environments for additional options:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community

Join the Ultralytics community to connect with others working on exciting projects like yours:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)
- Share and explore ideas on our [Subreddit](https://reddit.com/r/Ultralytics)

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing, ensuring compatibility and stability across platforms.

This is an automated response 🤖, but an Ultralytics engineer will review and assist you further soon. Thank you for sharing your work and contributing to the community! 🚀

**Comment by glenn-jocher** [comment]
@kyuhyong thanks for sharing your Jetson Nano deployment experience and Docker image! For official Ultralytics Jetson support, we recommend using our pre-built Docker images with TensorRT optimization as shown in the NVIDIA Jetson deployment guide. These include CUDA, PyTorch, and all dependencies preconfigured for different JetPack versions. Community contributions are always appreciated - we encourage users to explore both official and community resources.

**Reply by kyuhyong** [reply]
OK I downloaded the pre-built image that runs on my jetson nano and found slight differences.

First of all, can you share how did you manage to install python3.8 alongside python2.7?
I build my own image with below instruction to install 3.8 but found there is no python2.7 available. 
```
FROM nvcr.io/nvidia/l4t-base:r32.7.1
ENV VIRTUAL_ENV=/opt/venv
ENV PATH="$VIRTUAL_ENV/bin:$PATH"
# Intall python 3.8
RUN apt-get -y install --no-install-recommends \
    python3.8 \
    python3.8-dev \
    python3.8-venv \
    && curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py \
    && python3.8 get-pip.py \
    && python3.8 -m venv $VIRTUAL_ENV \
    && . $VIRTUAL_ENV/bin/activate \
    && python -m pip install --upgrade pip \
    && pip install wheel setuptools
RUN . $VIRTUAL_ENV/bin/activate \
```
Performancewise there is no noticeable difference running both yolov5nu.pt and yolov5nu.engine. 
Thanks,

**Reply by glenn-jocher** [reply]
@kyuhyong thanks for sharing your Jetson Nano setup! For Python 3.8 installations, we recommend following the official NVIDIA Jetson deployment guide which includes pre-configured Docker images with CUDA and TensorRT optimization. Python 2.7 is deprecated and not included in modern images - we strongly advise using Python 3.8+ for all Ultralytics YOLO deployments. The Ultralytics Jetson documentation provides detailed environment setup instructions.
